reps/2024-05-21-ray-kubectl-plugin.md

```diff
@@ -0,0 +1,95 @@
+
+## Motivation
+
+Today it is incredibly challenging for data scientists and AI researchers unfamiliar with Kubernetes to start using Ray on Kubernetes.
+However, running Ray on Kubernetes is advantageous or necessary in certain environments. These users struggle with KubeRay for a variety of reasons:
+1. **Unfamiliarity with Kubernetes API and Best Practices**: Users new to Kubernetes may struggle to operate Ray clusters using Kubernetes concepts like Pods and Services due to unfamiliarity with the Kubernetes API and best practices.
+2. **Complex Kubernetes Networking**: Kubernetes networking can be daunting for beginners. Understanding how to use kubectl port-forward or externally exposed Services to connect to Ray clusters can be challenging without prior experience.
+3. **Construction of Kubernetes YAML Manifests**: Creating YAML manifests for RayCluster, RayJob, and RayService can be challenging for those not experienced with using Kubernetes custom resources.
+
+For novice Kubernetes users, an intuitive CLI experience can drastically enhance their onboarding journey. A kubectl plugin is proposed since it can seamlessly integrate with the rest of the existing kubectl surface as needed.
+The primary goal of this plugin should be user-friendliness and a smooth progression from zero to “something good enough”. For more advanced scenarios and scaling requirements, users can opt to manage KubeRay custom resources
+independently as many users do today.
+
+More simply, the Ray Kubectl Plugin simplifies the management of Ray clusters by eliminating the need for users to handle complex YAML configurations.
+
+Why a kubectl plugin instead of the existing kuberay CLI?
+* **Convenience / ease of use**: Kubectl plugins are designed to work within the existing kubectl framework. They inherit the user's current authentication, context, and cluster metadata, eliminating the need for additional setup.
+* **Seamless kubectl integration**: Users can effortlessly switch between basic kubectl subcommands and the Ray plugin using a single command-line tool.
+* **Accessibility**: The Kuberay CLI needs the KubeRay API server to function. However, the majority of clusters using KubeRay don't run the KubeRay API server, which limits the accessibility of the CLI.
+
+## User Journey
+
+### Create / Manage a Ray cluster
+
+```
+$ kubectl ray cluster create my-cluster --ray-version=2.9.3
+    --worker-replicas=10 --worker-cpus=8 -worker-memory=32GB --worker-gpus=2
+```
+
+```
+kubectl ray cluster scale default-worker-group --cluster my-cluster --replicas=10
+```
+
+```
+kubectl ray cluster worker-group create cpu-pool --cluster my-cluster -worker-replicas=10
+    --worker-cpus=8 -worker-memory=32GB
+
+```
+
+```
+kubectl ray cluster delete my-cluster
+```
+
+### Ray Job Submissions (using RayJob)
+
+```
+$ kubectl ray job submit --cluster my-cluster --image=image:tag -- python job.py
```

Comment:
Would it be a pre-requisite for `my-cluster` to pre-exist before job submission? Or could this command create the cluster if not available, then submit?

Reply:
> Would it be a pre-requisite for my-cluster to pre-exist before job submission?

Currently, yes.

> Or could this command create the cluster if not available, then submit?

This is planned for future work. The MVP only includes reading operations for the Kubernetes API server; the writing operations are not included in the MVP.

Reply:
My intention for the `kubectl ray job submit` command is that it will create a RayJob. If you don't speify `--cluster`, the RayJob will be configured to create a new RayCluster for the job.

Reply:
As @kevin85421 said this is not in MVP scope and will be developed after the initial `kubectl ray cluster` functionality

Reply:
Similar requests from Slack

https://ray-distributed.slack.com/archives/C02GFQ82JPM/p1717058981028339